### PR TITLE
Enhanced table panel

### PR DESCRIPTION
**Pull request by madshall**
*_Features: *_
- Table transform merges series by intersecting columns or "Time" if there are no intersecting columns (a kind of LEFT INNER JOIN)
- Column can be hidden
- Column maximum width can be set
- Header can be hidden
- Coloring applied to empty("-") values too
- Search panel added / hidden by default
- Column can be renamed
- Clickable rows with URL parameters taken from row values
- Text alignment within column
- Coloring fixed: wrong color index
- Number decimals minimum value added and "onchange" rerenders the panel
- Unit "locale" added. Example: turns 123456789 into 123,456,789.00

### CODE REVIEW

**Comment by torkelo** [comment]
Some of this looks really great! But it's a bit sad that everything is bundled in one PR, makes it hard to review and merge individual features. 

Is there a change you can submit PR's for the individual features?

**Comment by madshall** [comment]
Unfortunately some of the features were fixed/improved in next commits so they overlap. I didn't make any changes out of table panel scope except the new unit in kbn.js

**Comment by torkelo** [comment]
Then I am not sure how to continue with this PR since it contains so many features, features that requires individual review, discussion, unit tests, decision if it's something that should be merged etc. 

Now it's all or nothing, or a lot of work on our part to extract some of the more straight forward features (like the locale unit, which does not have anything to do with table panel)

**Comment by madshall** [comment]
What can I do to make the merge process easier? If I create, let's say 10 branches, move each feature to its branch and make pull requests from each branch separately - will it help?

**Comment by torkelo** [comment]
yes, that will help a lot

**Comment by madshall** [comment]
Is there any chance this can be merged all together? It's really painful to split this into separate PRs...

**Comment by torkelo** [comment]
It could potentially be merged all at once but it could take a lot of time as each feature in the PR needs separate review, discussing if it's a feature that many would find useful, if unit tests are needed, if there is a better way to implement it, etc.

**Comment by jbogarin** [comment]
I am really interested in this. How can I help?

**Comment by torkelo** [comment]
could you describe what you mean by the first feature:

> Table transform merges series by intersecting columns or "Time" if there are no intersecting columns (a kind of LEFT INNER JOIN)

How is this different from what is done now?

**Comment by madshall** [comment]
@torkelo This is an automatic merge of several queries results. If you have a single query then it's showing fine as a table, however, if you add more queries they are not showing up at all. This feature shows them too and the resulting data is a set of series joined by same columns. 

An example:
Query A returns a row of data: time, field1, field2, tag1, tag2
Query B returns a row of data: time, field3, field4, tag3, tag2
The result: time, field1, field2, field3, field4, tag1, tag2, tag3 (assuming the tag2 values are the same, the data is joined by tag2)

Another example:
Query A returns: time, field1
Query B returns time, field2
The result: time, field1, field2 (since the timestamps are equal the data is joined by time)

So basically this feature allows to join data from multiple measurements into one table when it's impossible to get the data with a single query.

**Comment by torkelo** [comment]
@madshall so do you specify which column it should join on (time, tag2, etc)

**Comment by madshall** [comment]
No, you don't need to specify. If common tags found then the user grouped by them for purpose and series will be obviously merged by them, if not, then time will be used.

**Comment by torkelo** [comment]
so it joins by any matched column? sounds like it could cause very inconsistent or random results

**Comment by madshall** [comment]
it matches columns only if all values from column C1 in series A are a full or a sub-set of values from column C1 in series B. Just exactly like INNER JOIN in SQL works, no surprises or inconsistencies. The only difference is that in SQL you have to explicitly say what column to use for joining and here it finds the best matching column automatically, or uses time column if nothing was found.

**Comment by bmundt** [comment]
+1

**Comment by torkelo** [comment]
> The only difference is that in SQL you have to explicitly say what column to use for joining and here it finds the best matching column automatically, or uses time column if nothing was found.

What is the `best matching column`?  what if 2 tags match and one don't? what if all tags match but timestamp doesn't? I cannot seen any unit tests for the join feature so having hard time understanding how it supposed to work.

**Comment by bmundt** [comment]
I like some of the features listed and would like to at least have those (mainly the clickable rows) even if  this doesn't get approved because of joins.

**Comment by madshall** [comment]
If two tags match and the third one doesn't then the data will be merged by matching tags only, the third tag and time will not be used and will appear as regular columns. The point here is to attach additional columns to existing series when you can't access them with a single query. This feature doesn't do miracles with data itself, but it helps to represent that data in the way SQL-developers used to. It was mostly done because of the lack of JOIN functionality in timeseries dbs.

Here's a good real life example:
- We collect configuration data from multiple applications at their start-up and this data is never emitted after that. One of those configuration settings is app's PID.
- Later we receive heartbeats from these apps and we need to show this new data along with app PID that was stored earlier and its timestamp is different from heartbeat's timestamp.
- At this point we can't merge data by a single query.
- Here we use this "merged" table feature to attach earlier saved PID to the rest of columns.

File: public/app/plugins/panel/table/transformers.ts [hunk 1]
```diff
@@ -220,11 +220,15 @@ transformers['json'] = {
 };
 
 function applyAliasing(panel, model){
+  var hash = [];
   for (var i = 0; i < model.columns.length; i++) {
     for (var j = 0; j < panel.styles.length; j++) {
       var regex = kbn.stringToJsRegex(panel.styles[j].pattern);
-      if (panel.styles[j].alias && model.columns[i].text.match(regex)) {
-        model.columns[i].alias = model.columns[i].text.replace(regex, panel.styles[j].alias);
+      if (model.columns[i].text.match(regex)) {
```
Comment (Kotove) [inline public/app/plugins/panel/table/transformers.ts]: If there aren't any styles, column names doesn't displaying. Because this IF is not true "if (model.columns[i].text.match(regex)) {"
and col.alias will not be set at all.

**Comment by torkelo** [comment]
Sorry, closing this as it was never broken up into separate PRs and now drifted too far from master
